### DISCUSSION TITLE
How to deal with the dependencies in dependencies within TCA

### DISCUSSION THREAD

**Discussion by avdwerff** [pointfreeco/swift-composable-architecture#1343] [Q&A]
HI! 

I have 2 questions regarding how to deal with dependencies within the TCA.
So  have a simple client like this:

public struct AuthenticationClient: Sendable {
    public var login: @Sendable (LoginRequest) async throws -> Bool
    
    public init(
        login: @escaping @Sendable (LoginRequest) async throws -> Bool
    ) {
        self.login = login
    }
}

And then with a static instance:

public extension AuthenticationClient {
    static let `default` = Self(
        login: { request in
            
            let usersSvc: UsersService = AirHubSDK.userService()
            let result = try await usersSvc.login(email: request.email, password: request.password)
            
            await AuthDataHolderClient.shared.update(authData: .init(
                userId: result.userId,
                accessToken: result.token,
                refreshToken: result.refreshToken
            ))
            
            return await AirHubSDK.isAuthorized
            
        }
    )
}

What would be a good way to adjust the default impl. So that the userservice (from the AirHub SDK) is mockable and the same for the AuthDataHolderClient with respect to the TCA pattern. Making the user service static on the AuthenticationClient causes a warning of a non sendable property on a Sendable struct...

**Comment by corysullivan** [comment]
I'm unsure of the relationship between `AirHubSDK` and `AuthDataHolderClient`. If they are not directly related, I suggest creating two separate dependencies. At a minimum, I would make separate calls in your `AuthenticationClient` interface for each function. And then pull them together in your reducer.

```swift
struct AuthenticationClient {
  public var userInfo: @Sendable (LoginRequest) async throws -> UserInfo
  public var update: @Sendable (UserInfo) async throws -> Void
  public var isAuthorized: @Sendable () async throws -> Bool
}
``` 

Your live implementation would look something like this
```swift
extension AuthenticationClient {
  static let live = Self(
    userInfo: { request in
      let usersSvc: UsersService = AirHubSDK.userService()
      return try await usersSvc.login(email: request.email, password: request.password)
  }, update: { userInfo in
    await AuthDataHolderClient.shared.update(authData: .init(
                userId: userInfo.userId,
                accessToken: userInfo.token,
                refreshToken: userInfo.refreshToken
            ))
  }, isAuthorized: {
      await AirHubSDK.isAuthorized
  })
}
```

and a mocked version could be 

```swift
extension AuthenticationClient {
  static let mock = Self(
    userInfo: { request in
     return UserInfo(userId: "123", accessToken: "AAA", refreshToken: "111")
  }, update: { userInfo in
    return ()
  }, isAuthorized: {
      return true
  })
}
```

**Reply by avdwerff** [reply]
Thanks for your answer @corysullivan, I see what you mean. So instead of having 1 func doing everything, you would rather have more 'smaller' clients/dependancies and call them separately. But what if you really need more dependencies within a dependency..?

**Reply by corysullivan** [reply]
You can inject dependencies into other dependencies. See https://github.com/pointfreeco/isowords/blob/92bc4f8d02562be915fed0fd55d8eac6451ff8aa/App/iOS/App.swift#L100

The `serverConfig` depends on both the `apiClient` and the `build`
